src/translate.py

```diff
@@ -3693,6 +3958,13 @@ def process_instr(instr: Instruction) -> None:
             assert branch_condition is None
             branch_condition = arch.instrs_branches[mnemonic](args)
 
+        elif mnemonic in arch.instrs_decctr_branches:
+            assert branch_condition is None
+            branch_condition = arch.instrs_decctr_branches[mnemonic](args)
+
+            ctr = Register("ctr")
+            set_reg(ctr, BinaryOp.int(args.regs[ctr], "-", Literal(1)))
```

Comment:
I believe the decrement happens before the comparison. Maybe we should add an asm pattern that splits this into two fictive instructions (decrement + conditional branch) to reduce the number of instruction categories?

Reply:
I did this refactor, but it did end up making the output a bit worse...

And despite the ordering bug here, most of the C output was "correct" before (I think it could be from a bug from the way that phi assignments happen before `branch_condition`s are evaluated?)

Reply:
I see... The worse output is unfortunate, albeit in line with how regular `for (i = 0; i < N; i++)` output also tends to look (`do { ... temp = phi + 1; phi = temp; } while (temp < lim);`). It would be nice to improve that if we can figure out a good way. If it's a big worsening we *could* consider keeping the incorrect version, e.g. by changing the asm pattern to `move $temp, $ctr`, `addi $ctr, $ctr, -1`, `<branch> $temp`, though it'd be a bit uncomfortable. I wonder if there's cases where that gives buggy output in practice, e.g. non-loop occurrences of $ctr.

And yeah, phi assignments are treated as happening all at once at the end of the block, past branch_condition's, which isn't possible to represent in C. That's definitely something that would be nice to fix, it's tripped people up in the past, but it's not obvious how. Would need temps for phis or some more involved refactoring of the whole phi/temp system. Or some wallpapering hacks for the most common cases.

Reply:
Yeah - I like that if we decide we want to go back to the incorrect version, it's possible to do so by modifying `BranchCtrPatter` (we don't need to bring back `instrs_branch_decctr`).